### ISSUE TITLE
"Bad file descriptor"?

### ISSUE THREAD

**Issue by Elronnd** [darlinghq/darling#254]
I'm installing xcode, as described in the README, but I get a bunch of messages like this:

    cp: /Volumes/Xcode_7.3.1/Xcode.app/Contents/Developer/Documentation/DocSets/com.apple.adc.documentation.OSX.docset/Contents/Resources/Tokens/Objective-C/instp/NSLengthFormatter/unitStyle.xml: could not copy extended attributes to /Applications/Xcode.app/Contents/Developer/Documentation/DocSets/com.apple.adc.documentation.OSX.docset/Contents/Resources/Tokens/Objective-C/instp/NSLengthFormatter/unitStyle.xml: Bad file descriptor

`/Applications/Xcode.app/Contents/Developer/Documentation/DocSets/com.apple.adc.documentation.OSX.docset/Contents/Resources/Tokens/Objective-C/instp/NSLengthFormatter/unitStyle.xml` is created, and does have reasonable contents.  What's going on?

**Comment by ahyattdev** [comment]
Extended Attributes are an HFS+ feature that Darling does not currently support because everything runs on the host's filesystem. 

I am not sure why Xcode uses extended attributes here. It probably is meaningless. Anyway, no effect can be noticed now because the docsets can't be viewed until Xcode itself is working 😆.

**Comment by ahyattdev** [comment]
[This is how Netatalk deals with HFS+ Extended Attributes]http://netatalk.sourceforge.net/2.1/htmldocs/AppleVolumes.default.5.html)

**Comment by LubosD** [comment]
When copying extattrs fails, `cp` tries to log that, which fails with EBADF (as we don't have syslogd) and then prints out the wrong error message to stderr.

**Comment by DuIslingr** [comment]
Could AFS cause issues with darling in the future? iOS 10.3 seems to have it and I imagine it will be present in all future mac versions. Asking here since its somewhat relevant I think.

**Comment by LubosD** [comment]
@DuIslingr No, I think AFS will be relevant only to darling-dmg (aka hdiutil).

**Comment by bugaevc** [comment]
It's APFS, actually

**Comment by ylluminate** [comment]
I think @lundman could have some useful insights on this area with his dealings with Extended Attribute handling on ZFS for macOS.

@lundman, what have been your experiences in this area as far as file system support goes on macOS and do you have any thoughts in going in the reverse direction here since the goal is being able to store these attributes on non-mac filesystems such as EXT3/4, ZFS, or any Linux FS?

**Comment by lundman** [comment]
ext3/zfs already does xattrs (in style of forks), so unless you are going through a network filesystem to confuse things, you should be able to store them fine. One thing that differs between Linux/Unix and OsX is that OsX defines new vnops to handle xattrs, `vnop_getxattr` `vnop_setxattr` `vnop_removexattr` `vnop_listxattr` and corresponding libc syscalls. You'd have to map those to the openat style process.

Unless I completely misunderstood the question and answered something way deeper :)

**Comment by r-wilbur** [comment]
@ylluminate @lundman Did you folks make some progress towards supporting the extended attributes in the last 3 years?  What is left to be done?  Is there something I can do to help?

I'm getting a slightly different error message today.  I rebuilt from a fresh clone of source yesterday and installed it.  Today, I downloaded the OsX application[1] I hoped to run under darling as a .dmg[2], mounted it with
Darling [~]$ hdiutil attach Downloads/macOS\ Mojave\ Patcher.dmg
Darling [~]$ cp -r /Volumes/macOS\ Mojave\ Patcher/macOS\ Mojave\ Patcher.app /Applications/
[for each file I get a message of the following form]
cp: /Volumes/macOS Mojave Patcher/macOS Mojave Patcher.app/Contents/Resources/Quartz.framework/Frameworks/QuartzFilters.framework/Resources/sv.lproj/QFilterView.strings: could not copy extended attributes to /Applications/macOS Mojave Patcher.app/Contents/Resources/Quartz.framework/Frameworks/QuartzFilters.framework/Resources/sv.lproj/QFilterView.strings: Operation not supported on socket

References:
[1]  http://dosdude1.com/mojave/
[2]  https://gateway.pinata.cloud/ipfs/Qmds4TpDcAXszxMMACdBYK7XA8ih2kBoiuV2ZWejebgzgY/macOS%20Mojave%20Patcher.dmg

**Comment by LubosD** [comment]
This is of no concern - it's not a fatal error, just a warning.

**Comment by r-wilbur** [comment]
@LubosD What are the 'extended attributes' that APFS uses but Darling can't presently manage?

My application for Darling is to try and create a bootable USB memory stick containing an installer for macOS Mojave (10.14) and a program to patch that installation.  I have formatted a partition on the memory stick as OS X Extended (Journaled) (using a friend's mac, before the arrival of the pandemic), but now am interested in running a macOS program[*] under Darling (on my Linux computer) which, if successful, will make a bootable image on that partition containing the macOS Mojave installer and a program to patch the Mojave installation so that it will boot on my late-2009 mac mini.

If I can help get Darling to the place where this works, it seems to me that Darling wins some new compatibility and functionality and I win when I can patch my mac mini to boot macOS Mojave so I can run XCode 11+ for iOS development.

(Does this possibly belong under a different bug/issue with the title 'Support APFS Extended Attributes'?)
Reference:
[*]  http://dosdude1.com/mojave/#instructions

**Comment by r-wilbur** [comment]
In my understanding this involves copying files from darling filesystem to the USB memory stick partition (APFS) and getting the attributes correct so that the mac mini bootROM will recognize and boot from that partition.  Also so it is valid under macOS Mojave, which is what I believe it will boot into.

It looks like there is an invocation of 'chmod' by the application which is currently unsupported in darling as I get the error message:
usage:	chmod [-fhv] [-R [-H | -L | -P]] [-a | +a | =a  [i][# [ n]]] mode|entry file ...
	chmod [-fhv] [-R [-H | -L | -P]] [-E | -C | -N | -i | -I] file ...

Running the application also involves some GUI support, part of which is yet to be implemented in darling.
[macOS_Mojave_Patcher_notes.txt](https://github.com/darlinghq/darling/files/4656988/macOS_Mojave_Patcher_notes.txt)
[macOS_Mojave_Patcher.log](https://github.com/darlinghq/darling/files/4656989/macOS_Mojave_Patcher.log)

**Comment by r-wilbur** [comment]
@LubosD et al:  This is the only darling issue listed on the github page for folks interested in making their first contribution to darling[*], to wit:
"
Contribute to darlinghq/darling

Make your first contribution to this repository by tackling one of the issues listed below.

Each issue displayed here is a "good first issue," selected for its relative approachability for first-time contributors.
"

Reference:
[*]  https://github.com/darlinghq/darling/contribute

**Comment by LubosD** [comment]
There's no APFS support in Darling. If we're talking about the source DMG, then it must be HFS+.

Darling does support xattrs as of now, this includes HFS-specific attributes such as `com.apple.FinderInfo`. The problem is the namespaces feature used on Linux, which places additional restrictions on xattr names. So all xattrs starting with `com.apple.` should be translated to `user.com.apple.` during setxattr.c/fsetxattr.c/removexattr.c and back again in listxattr.c/getxattr.c.

This is relatively easy to do, but since it commonly only affects "Finder info" (e.g. pixel coordinates where to show this file in Finder) or potentially the resource fork (which is used only with Carbon apps, which we don't support yet), it's not a big deal.

Whatever issues you're having, they're really not related to the errors you're mentioning.

**Comment by r-wilbur** [comment]
@LubosD:  I stand corrected!  The filesystem is indeed HFS+ on the USB memory stick and I wasn't clear on the equivalence between 'Mac OS Extended' and HFS+.  So it looks like if we can map the HFS+ xattrs from the DMG filesystem to Linux filesystem in darling and then map back out to HFS+ filesystem on the USB memory stick the xattrs should successfully survive the whiplash.

I implemented the translation in darling/src/kernel/emulation/linux/xattr/
new files:  translate_namespace.{c|h}
changed fgetxattr.c, flistxattr.c, fremovexattr.c, fsetxattr.c, getxattr.c, listxattr.c, removexattr.c, setxattr.c
I have rebuilt and am going to test to see how it works.

darling "com.apple.FinderInfo" -> linux "user.com.apple.FinderInfo",
linux "user.com.apple.FinderInfo" -> darling "com.apple.FinderInfo"

The translation amounts to darling "<xattr name>" -> linux "user.<xattr name>"
linux "user.<xattr name>" -> darling "<xattr name>"
linux "<xattr name not beginning with 'user.'>" -> darling "<xattr name not beginning with 'user.'>"
